contracts/KetherNFT.sol

```diff
@@ -0,0 +1,117 @@
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
+
+import "./IKetherHomepage.sol";
+
+
+// TODO: Name this something really cool
+contract Wrapper {
+  constructor(address target, bytes memory payload) {
+    (bool success,) = target.call(payload);
+    require(success, "Wrapper: target call failed");
+
+    selfdestruct(payable(target));
+  }
+}
+
+contract KetherNFT is ERC721URIStorage {
+  /// instance is the KetherHomepage contract that this wrapper interfaces with.
+  IKetherHomepage public instance;
+
+  /// metadataSigner provides oracle authenticity for tokenURI content.
+  address public metadataSigner;
+
+  // TODO: Do we want to emit events?
+
+  constructor(address _ketherContract, address _metadataSigner) ERC721("Thousand Ether Homepage Ad", "1KAD") {
+    instance = IKetherHomepage(_ketherContract);
+    metadataSigner = _metadataSigner;
+  }
+
+  function _wrapPayload(uint _idx) internal view returns (bytes memory) {
+    return abi.encodeWithSignature("setAdOwner(uint256,address)", _idx, address(this));
+  }
+
+  function _wrapCreation(uint _idx) internal view returns (bytes memory) {
+    return abi.encodePacked(
+      type(Wrapper).creationCode,
+      abi.encode(address(instance), _wrapPayload(_idx)));
+  }
+
+  function precompute(uint _idx, address _owner) public view returns (bytes32 salt, address predictedAddress) {
+    salt = sha256(abi.encodePacked(_owner)); // FIXME: This can be more gas-efficient? Also worth salting something random here like block number?
+
+    bytes memory bytecode = _wrapCreation(_idx);
+
+    bytes32 hash = keccak256(
+      abi.encodePacked(
+        bytes1(0xff),
+        address(this),
+        salt,
+        keccak256(bytecode)
+      )
+    );
+
+    predictedAddress = address(uint160(uint256(hash)));
+    return (salt, predictedAddress);
+  }
+
+  function _getAdOwner(uint _idx) internal view returns (address) {
+      (address owner,,,,,,,,,) = instance.ads(_idx);
+      return owner;
+  }
+
+  /// wrap mints an NFT if the ad unit's ownership has been transferred to the
+  /// precomputed escrow address.
+  function wrap(uint _idx, address _owner) external {
+    (bytes32 salt, address precomputedWrapper) = precompute(_idx, _owner);
+
+    require(_getAdOwner(_idx) == precomputedWrapper, "KetherNFT: owner needs to be the correct precommitted address");
+
+    // Wrapper completes the transfer escrow atomically and self-destructs.
+    new Wrapper{salt: salt}(address(instance), _wrapPayload(_idx));
+
+    require(_getAdOwner(_idx) == address(this), "KetherNFT: owner needs to be KetherNFT after wrap");
+    _mint(_owner, _idx);
+  }
+
+  function unwrap(uint _idx, address _newOwner) external {
+    require(ownerOf(_idx) == msg.sender, "KetherNFT: unwrap for sender that is not owner");
+
+
+    instance.setAdOwner(_idx, _newOwner);
+    _burn(_idx);
+  }
+
+  // FIXME: If we let people set their own tokenURI, do we need another NSFW flag option here?
```

Comment:
We could also wrap `tokenURI` getter to check if ad is marked as nsfw and return a placeholder uri instead?

Reply:
Mainly I'd like to avoid nft wallets/marketplaces banning kethernft just because someone made an obscene tokenURI.